library/ccm.c

```diff
@@ -46,31 +47,19 @@ void mbedtls_ccm_init(mbedtls_ccm_context *ctx)
 }
 
 int mbedtls_ccm_setkey(mbedtls_ccm_context *ctx,
-                       mbedtls_cipher_id_t cipher,
+                       mbedtls_cipher_id_t cipher_id,
                        const unsigned char *key,
                        unsigned int keybits)
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    const mbedtls_cipher_info_t *cipher_info;
 
-    cipher_info = mbedtls_cipher_info_from_values(cipher, keybits,
-                                                  MBEDTLS_MODE_ECB);
-    if (cipher_info == NULL) {
-        return MBEDTLS_ERR_CCM_BAD_INPUT;
-    }
-
-    if (mbedtls_cipher_info_get_block_size(cipher_info) != 16) {
-        return MBEDTLS_ERR_CCM_BAD_INPUT;
-    }
-
-    mbedtls_cipher_free(&ctx->cipher_ctx);
+    mbedtls_block_cipher_free(&ctx->block_cipher_ctx);
 
-    if ((ret = mbedtls_cipher_setup(&ctx->cipher_ctx, cipher_info)) != 0) {
+    if ((ret = mbedtls_block_cipher_setup(&ctx->block_cipher_ctx, cipher_id)) != 0) {
         return ret;
     }
 
-    if ((ret = mbedtls_cipher_setkey(&ctx->cipher_ctx, key, keybits,
-                                     MBEDTLS_ENCRYPT)) != 0) {
+    if ((ret = mbedtls_block_cipher_setkey(&ctx->block_cipher_ctx, key, keybits)) != 0) {
         return ret;
```

Comment:
We could return `MBEDTLS_ERR_CCM_BAD_INPUT` here and above in order to preserve existing behaviour.